File: apollo-runtime/src/main/java/com/apollographql/apollo/subscription/OperationServerMessage.java
package com.apollographql.apollo.subscription;

import com.apollographql.apollo.internal.json.BufferedSourceJsonReader;
import com.apollographql.apollo.internal.json.JsonReader;
import com.apollographql.apollo.internal.json.ResponseJsonStreamReader;

import java.io.IOException;
import java.util.Collections;
import java.util.Map;

import org.jetbrains.annotations.NotNull;

import okio.Buffer;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;
import static java.util.Collections.unmodifiableMap;

@SuppressWarnings("WeakerAccess")
public abstract class OperationServerMessage {
  static final String JSON_KEY_ID = "id";
  static final String JSON_KEY_TYPE = "type";
  static final String JSON_KEY_PAYLOAD = "payload";

  OperationServerMessage() {
  }

  @NotNull public static OperationServerMessage fromJsonString(@NotNull String json) {
    checkNotNull(json, "json == null");
    try {
      Buffer buffer = new Buffer();
      buffer.writeUtf8(json);
      return OperationServerMessage.readFromJson(new BufferedSourceJsonReader(buffer));
    } catch (Exception e) {
      return new Unsupported(json);
    }
  }

  private static OperationServerMessage readFromJson(@NotNull JsonReader reader) throws IOException {
    checkNotNull(reader, "reader == null");

    ResponseJsonStreamReader responseJsonStreamReader = new ResponseJsonStreamReader(reader);
    Map<String, Object> messageData = responseJsonStreamReader.toMap();
    String id = (String) messageData.get(JSON_KEY_ID);
    String type = (String) messageData.get(JSON_KEY_TYPE);
    switch (type) {
      case ConnectionError.TYPE:
        return new ConnectionError(messagePayload(messageData));

      case ConnectionAcknowledge.TYPE:
        return new ConnectionAcknowledge();

      case Data.TYPE:
        return new Data(id, messagePayload(messageData));

      case Error.TYPE:
        return new Error(id, messagePayload(messageData));

      case Complete.TYPE:
        return new Complete(id);

      case ConnectionKeepAlive.TYPE:
        return new ConnectionKeepAlive();

      default:
        throw new IOException("Unsupported message");
    }
  }

  @SuppressWarnings("unchecked")
  private static Map<String, Object> messagePayload(Map<String, Object> messageData) {
    Map<String, Object> messagePayload = (Map<String, Object>) messageData.get(JSON_KEY_PAYLOAD);
    return messagePayload == null ? Collections.<String, Object>emptyMap() : unmodifiableMap(messagePayload);
  }

  public static final class ConnectionError extends OperationServerMessage {
    public static final String TYPE = "connection_error";
    public final Map<String, Object> payload;

    public ConnectionError(Map<String, Object> payload) {
      this.payload = payload;
    }
  }

  public static final class ConnectionAcknowledge extends OperationServerMessage {
    public static final String TYPE = "connection_ack";

    public ConnectionAcknowledge() {
    }
  }

  public static final class Data extends OperationServerMessage {
    public static final String TYPE = "data";
    public final String id;
    public final Map<String, Object> payload;

    public Data(String id, Map<String, Object> payload) {
      this.id = id;
      this.payload = payload;
    }
  }

  public static final class Error extends OperationServerMessage {
    public static final String TYPE = "error";
    public final String id;
    public final Map<String, Object> payload;

    public Error(String id, Map<String, Object> payload) {
      this.id = id;
      this.payload = payload;
    }
  }

  public static final class Complete extends OperationServerMessage {
    public static final String TYPE = "complete";
    public final String id;

    public Complete(String id) {
      this.id = id;
    }
  }

  public static final class ConnectionKeepAlive extends OperationServerMessage {
    public static final String TYPE = "ka";
  }

  public static final class Unsupported extends OperationServerMessage {
    public final String rawMessage;

    public Unsupported(String rawMessage) {
      this.rawMessage = rawMessage;
    }
  }
}

File: apollo-runtime/src/main/java/com/apollographql/apollo/subscription/OperationServerMessage.kt
package com.apollographql.apollo.subscription

import okio.Buffer

sealed class OperationServerMessage {
  companion object {
    const val JSON_KEY_ID = "id"
    const val JSON_KEY_TYPE = "type"
    const val JSON_KEY_PAYLOAD = "payload"

    @Suppress("DeprecatedCallableAddReplaceWith")
    @Deprecated("This method is deprecated. Use an OperationMessageSerializer instead.")
    fun fromJsonString(json: String): OperationServerMessage =
        ApolloOperationMessageSerializer.readServerMessage(Buffer().writeUtf8(json))
  }

  class ConnectionError(@JvmField val payload: Map<String, Any?>) : OperationServerMessage() {
    override fun equals(other: Any?): Boolean = other is ConnectionError && other.payload == payload
    override fun hashCode(): Int = payload.hashCode()

    companion object {
      const val TYPE = "connection_error"
    }
  }

  class ConnectionAcknowledge : OperationServerMessage() {
    override fun equals(other: Any?): Boolean = other is ConnectionAcknowledge
    override fun hashCode(): Int = javaClass.hashCode()

    companion object {
      const val TYPE = "connection_ack"
    }
  }

  class Data(@JvmField val id: String?, @JvmField val payload: Map<String, Any?>) : OperationServerMessage() {
    override fun equals(other: Any?): Boolean = other is Data && other.id == id && other.payload == payload
    override fun hashCode(): Int = 31 * (id?.hashCode() ?: 0) + payload.hashCode()

    companion object {
      const val TYPE = "data"
    }
  }

  class Error(@JvmField val id: String?, @JvmField val payload: Map<String, Any?>) : OperationServerMessage() {
    override fun equals(other: Any?): Boolean = other is Error && other.id == id && other.payload == payload
    override fun hashCode(): Int = 31 * (id?.hashCode() ?: 0) + payload.hashCode()

    companion object {
      const val TYPE = "error"
    }
  }

  class Complete(@JvmField val id: String?) : OperationServerMessage() {
    override fun equals(other: Any?): Boolean = other is Complete && other.id == id
    override fun hashCode(): Int = id?.hashCode() ?: 0

    companion object {
      const val TYPE = "complete"
    }
  }

  class ConnectionKeepAlive : OperationServerMessage() {
    override fun equals(other: Any?): Boolean = other is ConnectionKeepAlive
    override fun hashCode(): Int = javaClass.hashCode()

    companion object {
      const val TYPE = "ka"
    }
  }

  class Unsupported(@JvmField val rawMessage: String) : OperationServerMessage() {
    override fun equals(other: Any?): Boolean = other is Unsupported && other.rawMessage == rawMessage
    override fun hashCode(): Int = rawMessage.hashCode()
  }
}